Display MIQ/Red Hat Automate domain current/available version in the UI.

h-kataria opened the pull request:
Added changes to Display MIQ/Red Hat Automate domain current/available versions in the UI if they both don't match.

https://bugzilla.redhat.com/show_bug.cgi?id=1283037


@mkanoor @gmcculloug please review, This PR should be merged after https://github.com/ManageIQ/manageiq/pull/5548 is merged..

gmcculloug commented:
@h-kataria My understanding was this message would be displayed when you first enter Automate -> Explorer and the root "Datastore" element is selected in the tree and would contain a flash message for each domain that was out-of-sync.

With the current approach the user needs to click on the domain before seeing the message.

gmcculloug commented on app/controllers/miq_ae_class_controller.rb:
```diff
@@ -197,6 +198,13 @@ def get_node_info(node)
+    version = @record.version
+    available_version = @record.available_version
+    return if version.nil? || available_version.nil?
+    @version_message = _("Currently running %s version, you have %s version available.") % [version, available_version] if version != available_version
```
This message should be displayed for all datastores at the same time so it will need to include the domain name.  Maybe something like this: "%s domain: Current version - %s, Available version - %s"

Example: `ManageIQ domain: Current version - 5.3, Available version - 5.4`

h-kataria replied:
@gmcculloug this message in only displayed when user has selected MIQ or Red Hat domain in the tree, message is not displayed on any other node so i am not sure if showing a Domain name is needed.

h-kataria replied:
@gmcculloug about your other comment about showing this message on Datastore node, i had initially thought that it should only be displayed on Root node of Automate explorer tree, but model methods that are added are instance methods, @mkanoor suggested to add message on Domain level, i can rework these changes if model method can be changed to return version info for all domains in an array.

gmcculloug replied:
There should be one method that gets passed the domain and checks the version information and returns either nil or the message string.  The message that contains the domain name can be used in both places.  (No need to maintain two different strings for this.)

Something like this:
```ruby
  def add_all_domains_version_message(domains)
    @version_messages = domains.collect do |domain|
      domain_version_message(domain)
    end.compact
  end

  def domain_version_message(domain)
    version = domain.version
    available_version = domain.available_version
    return if version.nil? || available_version.nil?
    return if version != available_version

    _("Currently running %s version, you have %s version available.") % [version, available_version]
  end
```
Then above where you currently call `add_domain_version_message` you would do this:
`@version_message = domain_version_message(@record) if @record.domain?`

mkanoor commented:
@gmcculloug @h-kataria 
I think we should have the messages at the domain as well as the datastore level.

h-kataria commented:
@gmcculloug added messages on Datastore node as well, please review.

![datastore_node_message](https://cloud.githubusercontent.com/assets/3450808/11352164/d139602c-9208-11e5-882d-7cdb62a48a81.png)

mkanoor commented:
@h-kataria 
:+1: 
The only comment I have is there was to be some way of indicating in the UI that the domain is out of sync similar to how we show disabled domains.

h-kataria commented:
@mkanoor i think we can check with UX team after this release to see if they have better ideas of representing such nodes in the tree, rather than adding * or striking out special nodes in the tree, we already have nodes grayed out and striked out and now adding an * in the tree might be too much.

gmcculloug commented on app/controllers/miq_ae_class_controller.rb:
```diff
@@ -188,7 +189,9 @@ def get_node_info(node)
-      @grid_data = User.current_tenant.visible_domains
+      visible_domains = User.current_tenant.visible_domains
+      @grid_data = visible_domains
+      add_all_domains_version_message(visible_domains)
```
The `visible_domains` variable does not server any purpose here, just pass `@grid_data` into the `add_all_domains_version_message` method.

gmcculloug commented on app/controllers/miq_ae_class_controller.rb:
```diff
@@ -197,6 +200,20 @@ def get_node_info(node)
+  def add_all_domains_version_message(domains)
+    @version_messages = []
+    domains.collect { |dom| [dom.name, dom.version, dom.available_version] }.each do |name, current, available|
+      @version_messages.push(_("%s domain: Current version - %s  Available version - %s") % [name, current, available]) if !current.nil? && !available.nil? && current != available
```
Can we add a `,` between "Current" and "Available" like I had below:
`ManageIQ domain: Current version - 5.3, Available version - 5.4`

mkanoor commented on app/controllers/miq_ae_class_controller.rb:
```diff
@@ -197,6 +200,20 @@ def get_node_info(node)
+    end
+  end
+
+  def add_domain_version_message
```
@h-kataria 
These methods could be combined into something like this

def domain_version_message(domain)
  version = domain.version
  available_version = domain.available_version
  return if version.nil? || available_version.nil?
  _("Currently running %s version, you have %s version available.") % [version, available_version] if version != available_version
end

def add_all_domains_version_message(domains)
  @version_messages = []
   domains.each do |dom|
     domain_version_message(dom).tap { |msg| @version_messages.push << msg if msg }
  end
end

h-kataria commented:
@mkanoor @gmcculloug addressed your comments,  please re-review

screenshots:

![screenshot from 2015-11-24 11 59 47](https://cloud.githubusercontent.com/assets/3450808/11374209/e6dc5084-92a2-11e5-8a6b-bc68daae15ea.png)

![screenshot from 2015-11-24 11 51 08](https://cloud.githubusercontent.com/assets/3450808/11374211/ead3b9d4-92a2-11e5-8b36-7997197886aa.png)

gmcculloug commented:
:+1:

gmcculloug commented:
@Fryguy This PR is saying that travis is pending but clicking through shows travis completed successfully.  Any thoughts?

Fryguy commented:
@gmcculloug I'm not really sure.  If you notice the coverage all fell by a lot, which implies that one of the suites didn't "finish" properly.  I'm wondering if we should just kick off the whole thing again.
